Revision: 34577

RobertVannn opened the pull request:
3291 数据量较大时 es 上数据不全

【BUG】【new】【story】【全文索引】1w条记录只同步了部分记录到es端

【重现步骤】 
1.创建普通表，并创建全文索引，插入包含全文索引字段的记录1w条 
2.执行全文检索，检查结果 
`db.cs2.cl.find({"":{$Text:{query:{match_all:{}}}}}) `
【预期结果】 
1.步骤2中，全文检索返回1w条记录 
【实际结果】 
1.步骤2中，全文检索实际只返回了4837条记录 
```
{ 
  "_id": { 
    "$oid": "5a94d0d18881a5205f0012e3" 
  }, 
  "a": 4835, 
  "b": 4835, 
  "test1": "test04835" 
} 
{ 
  "_id": { 
    "$oid": "5a94d0d18881a5205f0012e4" 
  }, 
  "a": 4836, 
  "b": 4836, 
  "test1": "test04836" 
} 
Return 4837 row(s). 
```
此时在es端执行查询，只返回了4838条记录 
```
GET sys_4144749405_group3/group3/_search 
{ 
  "query": { 
    "match_all": {}} 
    } 
  } 
} 

{ 
  "took": 3, 
  "timed_out": false, 
  "_shards": { 
    "total": 5, 
    "successful": 5, 
    "skipped": 0, 
    "failed": 0 
  }, 
  "hits": { 
    "total": 4838, 
    "max_score": 1, 
    "hits": [ 
      { 
        "_index": "sys_4144749405_group3", 
        "_type": "group3", 
        "_id": "5a94d0d18881a5205f000004", 
        "_score": 1, 
        "_source": { 
          "a": 4 
        } 
      }, 
```
【备注】 
非必现，对应的数据组主节点的适配器日志见附件

PhillipsPaull commented:
请处理

RobertVannn commented:
【原因分析】 
适配器的接收缓冲区溢出导致。 
该缓冲区用于接收从 ES 返回的操作结果或数据（记录的 _id），当前其大小使用的是固定大小 1MB，其自动扩展能力未完善，因此在返回结果较大时有问题。 
通常用于接收返回的一批 _id 等信息时足够使用，但在 _bulk 操作中，返回结果包含了 _bulk 操作中每一个具体操作的结果，因此当 _bulk 中包含了足够多的操作时，其返回结果的大小就有可能超过当前的固定缓冲区的大小，发生写越界，造成不可预知的后果。

RobertVannn commented:
【自验证】 
```
> db.createCS('cs').createCL('cl') 
localhost:11810.cs.cl 
Takes 0.249230s. 
> db.cs.cl.createIndex('idx', {a:"text", b:"text", test1:"text"}) 
Takes 0.151886s. 
> for(var j=0;j<10;j++){ var doc=[]; for(var i=0; i<10000;i++){doc.push({a:j*10000+i,b:j*10000+i,test1:"test"+j*10000+i});}db.cs.cl.insert(doc);} 
Takes 15.820190s. 
> for(var j=10;j<100;j++){ var doc=[]; for(var i=0; i<10000;i++){doc.push({a:j*10000+i,b:j*10000+i,test1:"test"+j*10000+i});}db.cs.cl.insert(doc);} 
Takes 135.511715s. 
```


此时在 ES 上查到的数据量正确 
```
{ 
  "count": 1000001, 
  "_shards": { 
    "total": 5, 
    "successful": 5, 
    "skipped": 0, 
    "failed": 0 
  } 
} 
```

使用全文检索功能进行查询，结果正常： 
```
> db.cs.cl.find({$or:[{"":{"$Text":{"query":{"match":{"a":9999}}}}}, {a:9998}]}) 
{ 
  "_id": { 
    "$oid": "5a98b0809a783e5f6d099cf0" 
  }, 
  "a": 9998, 
  "b": 9998, 
  "test1": "test09998" 
} 
{ 
  "_id": { 
    "$oid": "5a98b0809a783e5f6d099cf1" 
  }, 
  "a": 9999, 
  "b": 9999, 
  "test1": "test09999" 
} 
Return 2 row(s). 
Takes 1.778915s. 
```

RobertVannn commented:
【提交记录】 
Revision: 34583 

Date: 2018年3月2日 9:23:36 
Message: 
3291 数据量较大时 es 上数据不全 

Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttpDef.hpp 

Revision: 34578 

Date: 2018年3月1日 20:11:48 
Message: 
3291 数据量较大时 es 上数据不全 

Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESClt.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESClt.hpp 

Revision: 34577 

Date: 2018年3月1日 20:11:45 
Message: 
3291 数据量较大时 es 上数据不全 

Modified : /trunk/SequoiaDB/engine/tools/seadapter/seAdptIndexSession.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESClt.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttp.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttpDef.hpp 

检查项---------------------------------------------------------结论（NA/N/Y） 

修改代码是否全部提交SVN主干和分支版本 
且符合代码规范--------------------------------------------Y 
是否用例全部通过-----------------------------------------Y 
修改代码是否通过检视------------------------------------提交检视 
资料是否变更-----------------------------------------------NA 
接口是否变更-----------------------------------------------NA 
相关驱动是否变更或添加---------------------------------NA 
是否保持向前兼容------------------------------------------NA 
是否影响安装部署和监控---------------------------------NA 
是否对影响性能--------------------------------------------NA 
sdb shell帮助信息是否更新------------------------------NA

已修改，请检视

PhillipsPaull commented:
2. 当前设定的通过 rest 接口在 ES 上操作的超时时间为 3s，概率性出现超时，调整为 10s； 

1、 db的会话断开，适配器能立既感知 
2、 适配器和es的连接 keep alive（按db默认方式） 
3、 超时时间可配置，默认10s 

评估一下连接的处理

RobertVannn commented:
【检视意见修改】 
1. 适配器侧用于索引和搜索的会话都是基于当前的异步会话机制，当会话断开时，通过对应的 pmdAsyncMsgHandler 进行处理： 
（1）索引会话管理器会将当前的所有索引会话退出； 
（2）搜索会话管理器会将当前所有的搜索会话退出（会话退出时会清理 ES 上的 scroll 资源） 

2. keep alive 的设置涉及到两个方面： 
（1）socket 属性中的 keep alive 设置，全部使用默认值； 
（2）HTTP 请求中，由于使用 HTTP 1.1 协议，在连接时默认就是使用长连接（即开户 Keepalive），之前在请求中包含 Connection: keep-alive 没有必要，调整为在 keep alive 为 false 的情况下，在头部中包含 Connection: close，即告知服务端，响应完成后关闭连接。当前的使用场景中均为 keep alive 为 true。 

3. 在 sdbseadapter 的配置项中增加了操作超时一项： optimeout，默认值为10000（10s），可通过配置文件或命令行参数进行配置 
```
$ sdbseadapter --help 
Command options (display): 
  -h [ --help ] help 
  --version version 
  -c [ --confpath ] arg Configure file path 
  -v [ --diaglevel ] arg Diagnostic level,default:3,value range:[0-5] 
  --datanodehost arg Data node address 
  --datasvcname arg Data node service name or port 
  --searchenginehost arg Search engine address 
  --searchengineport arg Search engine service name or port 
  -t [ --optimeout ] arg Rest operation timeout in millisecond,default:10000,va 
                         lue range[0-3600000] 
```

RobertVannn commented:
【自验证】
1. 断连感知可通过观察适配器日志（打开 DEBUG 日志 ）： 
当有全文索引时，适配器上的索引线程会打印如下信息： 
```
2018-03-23-16.14.48.415065 Level:DEBUG 
PID:1358 TID:1368 
Function:_sendGetmoreReq Line:551 
File:SequoiaDB/engine/tools/seadapter/seAdptIndexSession.cpp 
Message: 
Send getmore request to data node successfully 
```

当与数据节点连接断开后，会话退出，线程释放，打印类似如下信息： 
```
2018-03-23-16.14.48.815895 Level:EVENT 
PID:1358 TID:1368 
Function:pmdEDUEntryPoint Line:2029 
File:SequoiaDB/engine/pmd/pmdEDUMgr.cpp 
Message: 
Terminating thread[1368] for EDU[ID:6, Type:SeIndexAgent, Name: Type:Indexer,NodeID:71000,TID:2] 

```
对于搜索会话，可通过如下方式： 
在创建了全文索引的集合中插入多于1W条记录，然后使用游标进行遍历，只做一次 next，然后停止数据节点，在开始和节点停止后，适配器中会分别打印搜索会话的创建和退出信息： 
```
2018-03-23-16.22.56.434955 Level:EVENT 
PID:2227 TID:2244 
Function:pmdEDUEntryPoint Line:1920 
File:SequoiaDB/engine/pmd/pmdEDUMgr.cpp 
Message: 
Start thread[2244] for EDU[ID:7, type:SeAgent, Name:] 

2018-03-23-16.23.24.757396 Level:EVENT 
PID:2227 TID:2244 
Function:pmdEDUEntryPoint Line:2029 
File:SequoiaDB/engine/pmd/pmdEDUMgr.cpp 
Message: 
Terminating thread[2244] for EDU[ID:7, Type:SeAgent, Name: Type:SEAdptAgent,NodeID:71000,TID:2185] 
```


2. 超时时间的配置可观察适配器的日志中的启动信息： 
```
2018-03-23-16.21.58.453882 Level:EVENT 
PID:2227 TID:2227 
Function:pmdThreadMainEntry Line:199 
File:SequoiaDB/engine/tools/seadapter/pmdSEAdptMain.cpp 
Message: 
Start sdbseadapter[Version: 2.9, Release: 24386, Build: 2016-08-15-11.31.51(Debug)]... 

2018-03-23-16.21.58.454041 Level:EVENT 
PID:2227 TID:2227 
Function:pmdThreadMainEntry Line:205 
File:SequoiaDB/engine/tools/seadapter/pmdSEAdptMain.cpp 
Message: 
All configs: 
datanodehost = 192.168.31.99 
datasvcname = 11830 
diaglevel = 5 
searchenginehost = 192.168.31.99 
searchengineport = 9200 
optimeout = 20000 
confpath = conf/seadpt/11830/ 
```

RobertVannn commented:
【提交记录】 
Revision: 34972 

Date: 2018年3月23日 16:36:27 
Message: 
3291 提供超时配置，调整 keepalive 

Modified : /trunk/SequoiaDB/engine/tools/seadapter/seAdptAgentSession.hpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/seAdptMgr.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/seAdptOptionsMgr.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/seAdptOptionsMgr.hpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESClt.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESClt.hpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESCltMgr.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilESCltMgr.hpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttp.cpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttp.hpp 
Modified : /trunk/SequoiaDB/engine/tools/seadapter/utilHttpDef.hpp 
Modified : /trunk/conf/samples/sdbseadapter.conf 

Revision: 34973 

Date: 2018年3月23日 16:43:35 
Message: 
3291 更新适配器配置，增加 optimeout 

Modified : /trunk/doc/src/document/database_management/text_search/sdbseadapter.md 

检查项---------------------------------------------------------结论（NA/N/Y） 

修改代码是否全部提交SVN主干和分支版本 
且符合代码规范--------------------------------------------Y 
是否用例全部通过-----------------------------------------Y 
修改代码是否通过检视------------------------------------提交检视 
资料是否变更-----------------------------------------------Y 
接口是否变更-----------------------------------------------NA 
相关驱动是否变更或添加---------------------------------NA 
是否保持向前兼容------------------------------------------NA 
是否影响安装部署和监控---------------------------------NA 
是否对影响性能--------------------------------------------NA 
sdb shell帮助信息是否更新------------------------------NA

检视意见已修改，请检视

PhillipsPaull commented:
请安排测试

PhillipsPaull commented:
请验证

OwenMichaell commented:
检视通过，请关闭
